### PR TITLE
Fix #403: Near-term RF tuning

### LINKED ISSUE #403: Near-term RF tuning: (1) DCV-bound at asset level, (2) allocate on dataset rank, (3) publishers get 2x
_(Note: this was originally in [df-private#35](https://github.com/oceanprotocol/df-private/issues/35). I copied it here, now that it's going to implementation. I didn't transfer because gh doesn't allow transfer from private -> public repos.)_

### Motivation

Issue [df-private#22](https://github.com/oceanprotocol/df-private/issues/22) is the top-level EPIC/issue for Data Farming Reward Function (RF) design. We use it to explore candidate ideas to fix various issues (stake whales, etc). It sometimes has sub-tasks for sub-issues like "publisher incentives" ([df-private#30](https://github.com/oceanprotocol/df-private/issues/30)).

We like to lump RF updates into groups for release, so that it's easier for our community to track.

Recent iterations have led to a group of three candidate changes. Call them (1)(2)(3). They're all pragmatic, and don't introduce new issues (eg Sybil).

Aim of this issue: implement the three candidate changes, and get it released

### TODOs

- [x] Implement the items into a PR [here](https://github.com/oceanprotocol/df-py/pull/404)
  - [x] Implement (1): DCV-bound rewards at asset level, rather than global. Details below.
  - [x] Implement (2): Rather than allocate to datasets on DCV, allocate according to rank. Details below.
  - [x] Implement (3): Whatever the publisher has staked on the asset that they published, when we calculate rewards for them, it's as if they have 2x the stake. Details below.
  - [x] For (2): do plots and give recommended params. **[RF Ranking Study](https://docs.google.com/presentation/d/1zZdWfywruMPt6r7vfl0nQD8Fgj4wkfuMslZcWJmy3GE/edit#slide=id.p)**
- [x] Agree on a release date with the team. (Best to wait until implementation is done)
- [x] Prepare draft content
  - [x] Write draft blog content on "why and how of RF updates"
  - [x] Write draft tweetstorm
- [x] At appropriate time, roll out changes
  - [x] Merge the PR 
  - [x] Publish draft "why and how of RF updates" into relevant DF blog
  - [x] Publish draft tweetstorm
- [x] Write github issue for getting (3) shipped. Includes changes to backend, to frontend, and blog posts content. (The PR merged as part of *this* issue gets most of the way for the backend. So just some final work left.)

### Details on (1): DCV-bound rewards at asset level

DCV-bound rewards at asset level, rather than global.

Background: In [df-py#357](https://github.com/oceanprotocol/df-py/issues/357) [[commit](https://github.com/oceanprotocol/df-py/commit/3816d66d9309439410dd5d6037e06d314c105996)] [[PR](https://github.com/oceanprotocol/df-py/pull/360)], we implemented "bound total rewards by total DCV * mult". It was rolled out in DF Round 9 (DF9). It's a key building block to stopping wash consume.

The idea here is: rather than bounding _total_ rewards by DCV (global), bound them at the _asset_ level. (Similar to how we bound to <= 125% APY at asset level.)

Updated macro algorithm for DF rewards distribution:
- As usual, distribute rewards pro-rata to each asset on its DCV.
- Then, for each asset:
  - Compute rewards
  - Further bound rewards to the asset by its DCV * mult; and by 125% APY

Benefits: 
- People can't rely on a single actor to do wash consume for everyone; instead they need to do it themselves. This catalyzes more engagement from more people
- More robust RF, because it's specific to each asset: generalizing on the previous point, it likely bound other undesirable behaviors that we haven't yet identified
- Conceptually simpler implementation, side-by-side with 125% bounding

### Details on (2): Allocate to datasets on rank

Rather than allocate to datasets on DCV, allocate according to rank.

Implementing this will raise Q's about how much % to give to rank 1, rank 2, etc. E.g. is it 10% to rank 1, 9% to rank 2, etc. There are many options here. Here's how we'll approach it:
- build it, and parameterize the problem
- make 2-3 proposals of parameter settings (with data/plots). Leverage heuristics like "give 99% of rewards to top 20 datasets" or "rank 1 dataset should get <= 10% reward"
- review with everyone, and select one

### Details on (3): 2x to publishers

Whatever the publisher has staked on the asset that they published, when we calculate rewards for them, it's as if they have 2x the stake." 

- Make it where: whatever the publisher has staked on the asset that they published, when we calculate rewards for them, it's as if they have 2x the stake.
- Pro: simple to implement, simple to explain
- Con: incentivizes fraudulent publishing + wash consume. This will go away when DCV_mult hits 0.03.

Note: this is Cand E in [df-private#30](https://github.com/oceanprotocol/df-private/issues/30)).

Open Q: is 2x the best? Or should we go higher, e.g. 5x to really drive incentive to publish?

### PR DESCRIPTION
**Pull request by trentmc**
Fix #403

Changes proposed in this PR: 

- (1: DCV-bound): done
- (2: rank): done
  - [RF Ranking study](https://docs.google.com/presentation/d/1zZdWfywruMPt6r7vfl0nQD8Fgj4wkfuMslZcWJmy3GE/edit#slide=id.p)
- (3: publisher): done, but turned off

### CODE REVIEW

**Comment by alexcos20** [comment]
Please do not merge before Thu morning, when DF18 rewards are distributed

**Comment by trentmc** [comment]
This has not yet been tested in df-sql or df-webapp flows. It won't happen this week because Roberto and Berkay are on vacation.

Please do not merge until that testing has happened.

Cc @idiom-bytes

**Comment by jamiehewitt15** [comment]
@trentmc do you know why the tests are failing on this?

**Comment by trentmc** [comment]
> @trentmc do you know why the tests are failing on this?

Tests are now passing. It's ready for review. cc @idiom-bytes

File: util/query.py
Comment (trentmc) [reply util/query.py:444]: Great Q. 

The extra rewards are only going to creators (publishers), as per specified.

I added owners because I felt it would be useful for the frontend to show both owners and creators. Showing owners will be a complement to creators, and to help the user distinguish that one is not the other. It doesn't need to be in this PR, but it "fell out" conveniently from the rest of the work, hence it's here.

Comment (alexcos20) [reply util/query.py:444]: wait, this breaks all ERC721 selling models.
Once the ownership of an erc721 has been transferred, all rights & rewards should be transferred to new owner  (all rights are being transferred now in erc721 template, so I don't see any reason to have rewards act differently)

Comment (alexcos20) [inline util/query.py:444]:
```suggestion
                owner
```

File: util/query.py
Comment (alexcos20) [inline util/query.py:479]:
```suggestion
            creator_addr = order["datatoken"]["owner"]["id"].lower()
```

File: util/query.py
Comment (alexcos20) [reply util/query.py:444]: see my code suggestions below, because anyway, one line of code was broken

File: util/calcrewards.py
Comment (idiom-bytes) [reply util/calcrewards.py:259]: Yes, that is the outline I communicated

I also wrote the remainder here.
https://app.zenhub.com/workspaces/ve--df-60ddbf267dd0b500137f3c3d/issues/gh/oceanprotocol/df-issues/143

File: util/query.py
Comment (idiom-bytes) [reply util/query.py:479]: I agree with this.

If we can move forward with this, then :+1:

File: util/constants.py
Comment (jamiehewitt15) [inline util/constants.py:14]: It's good to see that we have the option to change the type of ranking scale that's used. Or ranking could in fact be turned of if we decided to move in a different direction.

Comment (jamiehewitt15) [reply util/constants.py:14]: I think we need a clear success criteria so that we can objectively decide which ranking scale is best and how many assets should receive rewards.

**Review by alexcos20 — APPROVED** [review]

**Comment by trentmc** [comment]
@idiom-bytes : Alex has approved this PR. Can it be merged?

**Review by trizin — APPROVED** [review]
We should replace all "creator" => "owner" before merging apart from that looks fine to me.

**Pull request merged**
